privacyidea/static/components/token/controllers/tokenRolloverController.js

```diff
@@ -0,0 +1,104 @@
+ * License along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+myApp.controller("tokenRolloverController", ["$scope", "TokenFactory", "$timeout",
+                                           "$stateParams", "$state", "instanceUrl",
+                                           "versioningSuffixProvider",
+    function tokenRolloverController($scope, TokenFactory, $timeout,
+                                     $stateParams, $state, instanceUrl,
+                                     versioningSuffixProvider) {
+
+    $scope.tokenSerial = $stateParams.tokenSerial;
+    $scope.qrCodeWidth = "250";
+    $scope.fileVersionSuffix = versioningSuffixProvider.$get();
+
+    // scroll to the top of the page
+    document.body.scrollTop = document.documentElement.scrollTop = 0;
+
+    // define functions
+    $scope.get = function () {
+        TokenFactory.getTokenForSerial($scope.tokenSerial, function (data) {
+            $scope.token = data.result.value.tokens[0];
+        });
+    }
+
+    $scope.getAndRegenerateToken = function () {
+        TokenFactory.getTokenForSerial($scope.tokenSerial, function (data) {
+            $scope.token = data.result.value.tokens[0];
+            $scope.regenerateToken();
+        });
+    };
+
+    $scope.regenerateToken = function () {
+        var params = $scope.token;
+        params.type = params.tokentype
+        // force server key generation (HOTP, TOTP)
+        params.genkey = true;
+        // use 2stepinit if preferred
+        params["2stepinit"] = $scope.checkRight(params.type + "_2step=force");
+        if (!$scope.form) { $scope.form = {}; };
+        $scope.form["2stepinit"] = params["2stepinit"];
+        // enroll token using the current serial, type and options
+        TokenFactory.enroll({}, params, $scope.callback);
+    };
+
+    // enrollment callback supports totp, hotp, push, paper, tan, registration
+    $scope.callback = function (data) {
+        $scope.enrolledToken = data.detail;
+        if ($scope.enrolledToken.otps) {
+            var otps_count = Object.keys($scope.enrolledToken.otps).length;
+            $scope.otp_row_count = parseInt(otps_count/5 + 0.5);
+            $scope.otp_rows = Object.keys($scope.enrolledToken.otps).slice(0, $scope.otp_row_count);
+        }
+        if ($scope.enrolledToken.rollout_state === "clientwait") {
+            $scope.pollTokenInfo();
+        }
+        $('html,body').scrollTop(0);
+    };
+
+    $scope.pollTokenInfo = function () {
+        TokenFactory.getTokenForSerial($scope.enrolledToken.serial, function(data) {
+            $scope.enrolledToken.rollout_state = data.result.value.tokens[0].rollout_state;
+            // Poll the data after 2.5 seconds again
+            if ($scope.enrolledToken.rollout_state === "clientwait") {
+                $timeout($scope.pollTokenInfo, 2500);
+            }
+        })
+    };
+
+    $scope.sendClientPart = function () {
```

Comment:
this doesn't work yet?

Reply:
fixed in 90b60853d1020aa950c45f486257760931a1b290

Reply:
If the 2step enrollment is forced, this works but if it is only allowed, it doesn't. It just enrolls a new token. But of course we would need to request the 2step enrollment in the modal before...